nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java

```diff
@@ -384,6 +385,127 @@ public void instantiationInUnannotatedCode() {
+            "   // BUG: Diagnostic contains: Cannot assign from type",
+            "   return new A<String>();",
+            "  }",
+            "  static void method3(int a) {",
+            "   System.out.println(a);",
+            "  }",
+            "  static A<@Nullable String> method4() {",
+            "   return new A<@Nullable String>();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void genericFunctionReturnTypeNormalTree() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.jspecify.annotations.Nullable;",
+            "class Test {",
+            "  static class A<T extends @Nullable Object> { }",
+            "  static A<String> method1(A<@Nullable String> a) {",
+            "   // BUG: Diagnostic contains: Cannot assign from type",
+            "   return a;",
+            "  }",
+            "  static A<@Nullable String> method2(A<@Nullable String> a) {",
+            "   return a;",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void genericFunctionReturnTypeMultipleReturnStatementsIfElseBlock() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.jspecify.annotations.Nullable;",
+            "class Test {",
+            "  static class A<T extends @Nullable Object> { }",
+            "  static A<String> method1(A<@Nullable String> a, int num) {",
+            "   if (num % 2 == 0) {",
+            "    // BUG: Diagnostic contains: Cannot assign from type",
+            "     return a;",
+            "    } else {",
+            "     return new A<String>();",
+            "    }",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void genericFunctionReturnTypeMultipleReturnStatementsInsideLoop() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.jspecify.annotations.Nullable;",
+            "class Test {",
+            "  static class A<T extends @Nullable Object> { }",
+            "  static A<String> method1(A<@Nullable String> a, int num) {",
+            "   for(int i = 0; i < num; i++) {",
+            "     if (i % 13 == 0) {",
+            "      // BUG: Diagnostic contains: Cannot assign from type",
+            "       return a;",
+            "     }",
+            "    } ",
+            "    return new A<String>();",
+            "  }",
+            "  static A<String> method2(A<@Nullable String> a, int num) {",
+            "   while(num != 0) {",
+            "     if (num % 13 == 0) {",
+            "      // BUG: Diagnostic contains: Cannot assign from type",
+            "       return a;",
+            "     }",
+            "     num --;",
+            "    } ",
+            "    return new A<String>();",
+            "  }",
```

Comment:
Isn't this one redundant with `method1`?

Reply:
wanted to try while and for both

Reply:
But why would our code have anything to do with whether the returns are nested inside a loop?  To me this test seems a bit unnecessary

Reply:
ok. will remove it.